<a href="https://colab.research.google.com/github/oimartin/Older-and-Wiser/blob/geriatric_visits/geriatric_visits_nb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Load Data
Geriatric visits data: https://data.cms.gov/provider-data/dataset/bce0-b5db
<br>
Zip code/ State data: https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-statistics-2019-zip-code-data-soi

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px

In [2]:
visits = pd.read_csv("https://raw.githubusercontent.com/oimartin/Older-and-Wiser/geriatric_visits/data/Geriatric_Medicine.csv")
zip_codes = pd.read_csv('https://raw.githubusercontent.com/oimartin/Older-and-Wiser/geriatric_visits/data/all_states_only_zip_codes.csv')

# EDA

## Geriatric Visits

In [3]:
# First look at data
display(visits.info())
print(" ")
print(" ")

display(visits.describe())
print(" ")
print(" ")

# Only two codes used, new and existing patients
print("most_utilized_procedure_code_for_new_patient")
display(visits.iloc[:, 7].value_counts())
print("most_utilized_procedure_code_for_established_patient")
display(visits.iloc[:, 14].value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42959 entries, 0 to 42958
Data columns (total 15 columns):
 #   Column                                                Non-Null Count  Dtype  
---  ------                                                --------------  -----  
 0   zip_code                                              42959 non-null  int64  
 1   min_medicare_pricing_for_new_patient                  42959 non-null  float64
 2   max_medicare_pricing_for_new_patient                  42959 non-null  float64
 3   mode_medicare_pricing_for_new_patient                 42959 non-null  float64
 4   min_copay_for_new_patient                             42959 non-null  float64
 5   max_copay_for_new_patient                             42959 non-null  float64
 6   mode_copay_for_new_patient                            42959 non-null  float64
 7   most_utilized_procedure_code_for_new_patient          42959 non-null  int64  
 8   min_medicare_pricing_for_established_patient          42

None

zip_code  min_medicare_pricing_for_new_patient  \
count  42959.000000                          42959.000000   
mean   49591.379245                             58.238339   
std    27974.499582                              4.045550   
min      210.000000                             53.136000   
25%    26037.500000                             55.312000   
50%    48815.000000                             56.936000   
75%    73042.500000                             59.432000   
max    99950.000000                             74.816000   

       max_medicare_pricing_for_new_patient  \
count                          42959.000000   
mean                             177.239252   
std                               11.124559   
min                              163.672000   
25%                              169.736000   
50%                              174.056000   
75%                              180.408000   
max                              233.632000   

       mode_medicare_pricing_for_new_patient  min_copay_for_new_patient  \
count                           42959.000000               42959.000000   
mean                              177.239252                  14.559585   
std                                11.124559                   1.011387   
min                               163.672000                  13.284000   
25%                               169.736000                  13.828000   
50%                               174.056000                  14.234000   
75%                               180.408000                  14.858000   
max                               233.632000                  18.704000   

       max_copay_for_new_patient  mode_copay_for_new_patient  \
count               42959.000000                42959.000000   
mean                   44.309813                   44.309813   
std                     2.781140                    2.781140   
min                    40.918000                   40.918000   
25%                    42.434000                   42.434000   
50%                    43.514000                   43.514000   
75%                    45.102000                   45.102000   
max                    58.408000                   58.408000   

       most_utilized_procedure_code_for_new_patient  \
count                                       42959.0   
mean                                        99205.0   
std                                             0.0   
min                                         99205.0   
25%                                         99205.0   
50%                                         99205.0   
75%                                         99205.0   
max                                         99205.0   

       min_medicare_pricing_for_established_patient  \
count                                  42959.000000   
mean                                      18.073701   
std                                        1.459078   
min                                       16.256000   
25%                                       17.040000   
50%                                       17.616000   
75%                                       18.544000   
max                                       23.768000   

       max_medicare_pricing_for_established_patient  \
count                                  42959.000000   
mean                                     144.793623   
std                                        9.072558   
min                                      133.848000   
25%                                      138.800000   
50%                                      142.080000   
75%                                      147.312000   
max                                      190.744000   

       mode_medicare_pricing_for_established_patient  \
count                                   42959.000000   
mean                                      103.643492   
std                                         6.628021   
min                                        95.608000   
25%             

 
 
most_utilized_procedure_code_for_new_patient


99205    42959
Name: most_utilized_procedure_code_for_new_patient, dtype: int64

most_utilized_procedure_code_for_established_patient


99214    42959
Name: most_utilized_procedure_code_for_established_patient, dtype: int64

In [4]:
visits.nunique()

zip_code                                                42959
min_medicare_pricing_for_new_patient                       95
max_medicare_pricing_for_new_patient                       96
mode_medicare_pricing_for_new_patient                      96
min_copay_for_new_patient                                  95
max_copay_for_new_patient                                  96
mode_copay_for_new_patient                                 96
most_utilized_procedure_code_for_new_patient                1
min_medicare_pricing_for_established_patient               90
max_medicare_pricing_for_established_patient               96
mode_medicare_pricing_for_established_patient              96
min_copay_for_established_patient                          90
max_copay_for_established_patient                          96
mode_copay_for_established_patient                         96
most_utilized_procedure_code_for_established_patient        1
dtype: int64

<h1> Code Definitions </h1>
<b>99214</b>: Established patient office visit, 30-39 minutes
<br>
<b>99205</b>: Typically 60 minutes, Comprehensive history, Comprehensive Examination, High complexity medical decision making

## Zip Codes

In [5]:
# Observe zip code data from 2019 IRS
display(zip_codes.info())
print(" ")
print(" ")

display(zip_codes.nunique())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27695 entries, 0 to 27694
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   STATEFIPS  27695 non-null  int64 
 1   STATE      27695 non-null  object
 2   ZIPCODE    27695 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 649.2+ KB


None

STATEFIPS       51
STATE           51
ZIPCODE      27595
dtype: int64

In [6]:
# Select for only zipcodes, not overall state data
clean_zip = zip_codes[(zip_codes['ZIPCODE'] !=0) &
                      (zip_codes['ZIPCODE'] !=99999)]
display(clean_zip.info())
print(" ")
print(" ")
display(clean_zip.nunique())
display(clean_zip.describe())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27593 entries, 1 to 27693
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   STATEFIPS  27593 non-null  int64 
 1   STATE      27593 non-null  object
 2   ZIPCODE    27593 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 862.3+ KB


None

STATEFIPS       51
STATE           51
ZIPCODE      27593
dtype: int64

STATEFIPS       ZIPCODE
count  27593.000000  27593.000000
mean      29.669663  48856.241474
std       15.119769  27047.660765
min        1.000000   1001.000000
25%       18.000000  27105.000000
50%       29.000000  48845.000000
75%       42.000000  70578.000000
max       56.000000  99901.000000

## Add States and Regions to Visits

In [7]:
# functions to add states to visit df

# Determine start and end zip codes by state
def find_start_end_zip(current_state):
  zipcodes_state = clean_zip[clean_zip['STATE'] == state]['ZIPCODE']
  min_zip = zipcodes_state.min()
  max_zip = zipcodes_state.max()

  return min_zip, max_zip, current_state

# Add appropriate state to visit df by index position of zip code
def add_state(min_zip, max_zip, current_state):
  state_zip_index = visits[(visits['zip_code'] >= min_zip) & (visits['zip_code'] <= max_zip)].index
  visits.iloc[state_zip_index, -1] = current_state
  # print(f'{current_state} min zip code = {min_zip}, \n max zip code = {max_zip}')

  return 

In [8]:
# list of all state abbreviations
all_states_dc = list(clean_zip['STATE'].unique())

# create new column for states
visits['State'] = ''

# add states to visit data and print zip code boundaries by state
for state in all_states_dc:
  state_min, state_max, current_state = find_start_end_zip(state)
  add_state(state_min, state_max, current_state)


In [9]:
# Observe number of zipcodes labeled by state

fig = px.bar(visits['State'].value_counts(), title='Number of Zipcodes by State',
             labels={'value':'Zipcode Counts', 'index':'State'})
fig.update_layout(template="plotly_white",showlegend=False) 
fig.update_coloraxes(colorbar_orientation='h')
fig.show()

<h1>Zip Codes by State</h1>
The zip code data used is from the 2019 Individual income tax statistics by zip code. There are 559 zip codes that were not assigned a state.
2019 Individual income tax statistics by zip code. There are 559 zip codes that were not assigned a state.

## Adjust Zip_code in Visits

In [10]:
visits.iloc[:, 7] = visits.iloc[:, 7].astype('string')
visits.iloc[:, 14] = visits.iloc[:, 14].astype('string')

visits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42959 entries, 0 to 42958
Data columns (total 16 columns):
 #   Column                                                Non-Null Count  Dtype  
---  ------                                                --------------  -----  
 0   zip_code                                              42959 non-null  int64  
 1   min_medicare_pricing_for_new_patient                  42959 non-null  float64
 2   max_medicare_pricing_for_new_patient                  42959 non-null  float64
 3   mode_medicare_pricing_for_new_patient                 42959 non-null  float64
 4   min_copay_for_new_patient                             42959 non-null  float64
 5   max_copay_for_new_patient                             42959 non-null  float64
 6   mode_copay_for_new_patient                            42959 non-null  float64
 7   most_utilized_procedure_code_for_new_patient          42959 non-null  string 
 8   min_medicare_pricing_for_established_patient          42

## Add Regions

In [11]:
# Create regions
visits['Region'] = visits['State']
visits['Region'].replace(['CT','ME','MA','NH','RI', 'VT'],
                         'New England',inplace=True)

visits['Region'].replace(['NJ','NY','PA'],
                         'Middle Atlantic',inplace=True)

visits['Region'].replace(['IN','IL', 'MI','OH', 'WI'],
                         'East North Central',inplace=True)

visits['Region'].replace(['IA', 'NE', 'KS', 'ND','MN', 'SD', 'MO'],
                          'West North Central',inplace=True)

visits['Region'].replace(['DE', 'DC', 'FL', 'GA', 'MD', 'NC','SC', 'VA', 'WV'],
                          'South Atlantic',inplace=True)

visits['Region'].replace(['AL', 'LA', 'OK','TX'],
                          'West South Central',inplace=True)

visits['Region'].replace(['AR', 'KY', 'MS','TN'],
                          'East South Central',inplace=True)

visits['Region'].replace(['AZ', 'CO', 'ID','NM','MT', 'UT', 'NV', 'WY'],
                          'Mountain',inplace=True)

visits['Region'].replace(['AK', 'CA', 'HI', 'OR','WA'],
                          'Pacific',inplace=True)

visits.reset_index(drop=True, inplace=True)

In [12]:
# Observe number of zipcodes labeled by state
fig = px.bar(visits['Region'].value_counts(), title='Number of Zipcodes per Region',
             labels={'value':'Zipcode Counts', 'index':'Region'})
fig.update_layout(template="plotly_white",showlegend=False) 
fig.update_coloraxes(colorbar_orientation='h')
fig.show()

In [29]:
no_region_zip = visits[visits['Region'] == '']
bins = pd.IntervalIndex.from_tuples([(1, 10000),
                                     (10001, 20000),
                                     (20001, 30000),
                                     (30001, 40000),
                                     (40001, 50000),
                                     (50001, 60000),
                                     (60001, 70000),
                                     (70001, 80000),
                                     (80001, 90000),
                                     (90001, 100000)])

binned = pd.cut(no_region_zip['zip_code'].copy(), bins)
mod_bin = no_region_zip.rename(columns={'zip_code':'orig_zip_code'}).copy()
bin_new_visits = pd.concat([mod_bin, binned], axis='columns')


missing_region = bin_new_visits.groupby('zip_code').count().reset_index().iloc[:, 0:2]
missing_region.iloc[:,0] = missing_region.iloc[:,0].astype(str)
px.bar(missing_region, x='zip_code', y='orig_zip_code')

In [30]:
new_visits = visits[visits['Region'] != ''].copy()

# Graphs

## Treemap - Region

In [33]:
bins = pd.IntervalIndex.from_tuples([(95, 99.999999),
                                     (100, 104.999999),
                                     (105, 109.999999),
                                     (110, 114.999999),
                                     (115, 119.999999),
                                     (120, 124.999999),
                                     (125, 129.999999),
                                     (130, 134.999999),
                                     (135, 139.999999)])

binned = pd.cut(new_visits['mode_medicare_pricing_for_established_patient'].copy(), bins)
mod_bin = new_visits.rename(columns={'mode_medicare_pricing_for_established_patient':'orig_mode_med_estab'}).copy()
bin_new_visits = pd.concat([mod_bin, binned], axis='columns')
bin_new_visits.iloc[:,-1] = bin_new_visits.iloc[:,-1].astype(str).replace(to_replace={'(95.0, 99.999999]': '$95 to $99.9',
                                                                                      '(100.0, 104.999999]': '$100 to $104.99',
                                                                                      '(105.0, 109.999999]': '$105 to $109.99',
                                                                                      '(110.0, 114.999999]': '$110 to $114.99',
                                                                                      '(115.0, 119.999999]': '$115 to $119.99',
                                                                                      '(120.0, 124.999999]': '$120 to $124.99',
                                                                                      '(125.0, 129.999999]': '$125 to $129.99',
                                                                                      '(130.0, 134.999999]' : '$130 to $134.99',
                                                                                      '(135.0, 139.999999]': '$135 to $139.99'})
grouped_visits = bin_new_visits.groupby(['Region','State','mode_medicare_pricing_for_established_patient']).count().reset_index().iloc[:,0:4]


In [63]:
grouped_visits.head(40)

Region State mode_medicare_pricing_for_established_patient  \
0   East North Central    IL                               $100 to $104.99   
1   East North Central    IL                               $110 to $114.99   
2   East North Central    IN                                  $95 to $99.9   
3   East North Central    MI                               $100 to $104.99   
4   East North Central    MI                               $105 to $109.99   
5   East North Central    OH                               $100 to $104.99   
6   East North Central    WI                                  $95 to $99.9   
7   East South Central    AR                                  $95 to $99.9   
8   East South Central    KY                                  $95 to $99.9   
9   East South Central    MS                                  $95 to $99.9   
10  East South Central    TN                                  $95 to $99.9   
11     Middle Atlantic    NJ                               $110 to $114.99   
12     Middle Atlantic    NJ                               $115 to $119.99   
13     Middle Atlantic    NY                               $100 to $104.99   
14     Middle Atlantic    NY                               $110 to $114.99   
15     Middle Atlantic    NY                               $120 to $124.99   
16     Middle Atlantic    NY                               $125 to $129.99   
17     Middle Atlantic    PA                               $100 to $104.99   
18     Middle Atlantic    PA                               $110 to $114.99   
19            Mountain    AZ                               $100 to $104.99   
20            Mountain    CO                               $105 to $109.99   
21            Mountain    ID                                  $95 to $99.9   
22            Mountain    MT                               $100 to $104.99   
23            Mountain    NM                               $100 to $104.99   
24            Mountain    NV                               $105 to $109.99   
25            Mountain    UT                               $100 to $104.99   
26            Mountain    WY                               $100 to $104.99   
27            Mountain    WY                                  $95 to $99.9   
28         New England    CT                               $110 to $114.99   
29         New England    CT                               $120 to $124.99   
30         New England    MA                               $105 to $109.99   
31         New England    MA                               $115 to $119.99   
32         New England    ME                               $100 to $104.99   
33         New England    ME                                  $95 to $99.9   
34         New England    NH                               $105 to $109.99   
35         New England    RI                               $105 to $109.99   
36         New England    VT                               $100 to $104.99   
37         New England    VT                               $105 to $109.99   
38             Pacific    AK                               $135 to $139.99   
39             Pacific    CA                               $105 to $109.99   

    zip_code  
0       1246  
1        383  
2        999  
3        934  
4        249  
5       1483  
6        918  
7        728  
8       1030  
9        541  
10       813  
11       319  
12       423  
13      1232  
14       324  
15       602  
16        84  
17      1879  
18       363  
19       565  
20       675  
21       230  
22       411  
23       435  
24       255  
25       359  
26       198  
27       103  
28       433  
29         1  
30       491  
31       257  
32        96  
33       408  
34       284  
35        88  
36       308  
37         2  
38       264  
39      1207

In [59]:
colors = {'$95 to $99.9': '#f5fdfd',
                   '$100 to $104.99':'#e5ffff',
                   '$105 to $109.99':'#ccedea',
                   '$110 to $114.99':'#aad9d4',
                   '$115 to $119.99':'#4daf91',
                   '$120 to $124.99':'#5f9997',
                   '$125 to $129.99':'#315e58',
                   '$130 to $134.99':'#002a00',
                   '$135 to $139.99':'#0b1b05',
          '(?)':'lightgrey'}

In [60]:
fig = px.treemap(grouped_visits,
                 path=['Region','State', 'mode_medicare_pricing_for_established_patient'],
                 values='zip_code',
                 color='mode_medicare_pricing_for_established_patient',
                 color_discrete_map=colors,
                 width=1000, height=1000)

fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))

fig.update_coloraxes(colorbar_orientation='h')
fig.update_traces(marker_line_width = 0.3)
fig.show()

In [61]:
fig = px.treemap(grouped_visits[grouped_visits['Region'] == 'Middle Atlantic'],
                 path=['Region','State', 'mode_medicare_pricing_for_established_patient'],
                 values='zip_code',
                 color='mode_medicare_pricing_for_established_patient',
                 color_discrete_map=colors,
                 width=600, height=500)

fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))
fig.update_coloraxes(colorbar_orientation='h')
fig.update_traces(marker_line_width = 2)
fig.show()

In [62]:
fig = px.treemap(grouped_visits[grouped_visits['Region'] == 'East South Central'],
                 path=['Region','State', 'mode_medicare_pricing_for_established_patient'],
                 values='zip_code',
                 color='mode_medicare_pricing_for_established_patient',
                 color_discrete_map=colors,
                 width=600, height=500)

fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))
fig.update_coloraxes(colorbar_orientation='h')
fig.update_traces(marker_line_width = 2)
fig.show()